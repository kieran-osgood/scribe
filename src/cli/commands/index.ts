// export * from './base-command';
// export * from './default-command';
export * from './init-command.js';
