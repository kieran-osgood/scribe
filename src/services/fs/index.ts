// https://github.com/Effect-TS/node/pull/629/commits/dbbfc3b0f6f7e0b21bcca5bd095612ea998f49b6
export * from './error.js';
export * from './fs.js';
