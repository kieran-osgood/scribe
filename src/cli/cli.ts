import { Command } from '@effect/cli';

import packageJson from '../../package.json';
import { ScribeInit } from './commands';
import { ScribeDefault } from './commands/default-command';

const command = ScribeDefault.pipe(Command.withSubcommands([ScribeInit]));

export const run = Command.run(command, {
  name: 'Scribe',
  version: packageJson.version,
});
