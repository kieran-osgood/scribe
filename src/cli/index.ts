import { Effect } from '@scribe/core';
import { Builtins, Cli } from 'clipanion';
import { BaseContext } from 'clipanion/lib/advanced/Cli';
import { DefaultCommand } from './commands/DefaultCommand';

export function _Cli(args: string[], contextOverrides?: Partial<BaseContext>) {
  const cli = new Cli({
    binaryLabel: `scribe`,
    binaryName: 'scribe',
    enableCapture: true,
  });

  cli.register(DefaultCommand);
  // cli.register(InitCommand);

  cli.register(Builtins.HelpCommand);
  cli.register(Builtins.VersionCommand);
  return cli.runExit(args, { ...Cli.defaultContext, ...contextOverrides });
}

export function run(args: string[], contextOverrides: Partial<BaseContext>) {
  return Effect.tryPromise(() => _Cli(args.slice(2), contextOverrides));
}
