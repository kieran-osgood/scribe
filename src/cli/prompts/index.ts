import { Prompt } from '@effect/cli';
import { Effect } from 'effect';

export const fileName = Prompt.text({
  message: 'Name:',
  validate: s =>
    /^([A-Za-z\-_\d])+$/.test(s)
      ? Effect.succeed(s)
      : Effect.fail(
          'File name may only include letters, numbers & underscores.',
        ),
});

export const continueWarning = Prompt.toggle({
  message: 'Continue?',
  active: 'yes',
  inactive: 'no',
});

export const templates = (s: string[]) =>
  Prompt.select({
    message: 'Template:',
    choices: s.map(_ => ({
      title: _,
      value: _,
      // https://github.com/kieran-osgood/scribe/issues/32
      // description: 'This is some description of a template',
    })),
  });
