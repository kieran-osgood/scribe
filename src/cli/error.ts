import { TaggedClass } from 'src/core';

export class CliError extends TaggedClass('CliError')<{
  readonly cause?: unknown;
}> {}
