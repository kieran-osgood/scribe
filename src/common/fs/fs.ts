import { Effect } from '@scribe/core';
import * as NFS from 'fs';
import { ErrnoError } from './error';

export const writeFile = (
  path: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
) =>
  Effect.async<never, ErrnoError, boolean>(resume => {
    NFS.writeFile(path, data, options, err => {
      if (err) {
        resume(Effect.fail(new ErrnoError(err)));
      } else {
        resume(Effect.succeed(true));
      }
    });
  });
