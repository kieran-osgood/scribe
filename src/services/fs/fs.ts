import { Abortable } from 'node:events';

import { Context, Effect, pipe } from 'effect';
import * as NFS from 'fs';
import * as memfs from 'memfs';
import path from 'path';

import { MkDirError, ReadFileError, StatError, WriteFileError } from './error';

export interface FS {
  writeFile: typeof NFS.writeFile;
  readFile: typeof NFS.readFile;
  mkdir: typeof NFS.mkdir;
  stat: typeof NFS.stat;
  open: typeof NFS.open;
  close: typeof NFS.close;
}

export const FS = Context.Tag<FS>();
export const FSLive = NFS;
export const FSMock = memfs.fs as unknown as typeof NFS;

export const getFS = (test: boolean) => {
  return test ? FSMock : FSLive;
};

export const writeFile = (
  file: NFS.PathOrFileDescriptor,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, WriteFileError, NFS.PathOrFileDescriptor>(resume => {
        fs.writeFile(file, data, options, error => {
          if (error) {
            resume(
              Effect.fail(new WriteFileError({ file, data, options, error })),
            );
          } else {
            resume(Effect.succeed(file));
          }
        });
      }),
    ),
  );

export const writeFileWithDir = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions,
): Effect.Effect<FS, WriteFileError | MkDirError, NFS.PathOrFileDescriptor> =>
  pipe(
    mkdir(path.dirname(pathName), { recursive: true }),
    Effect.flatMap(() => writeFile(pathName, data, options)),
  );

export const readFile = (
  path: NFS.PathOrFileDescriptor,
  options:
    | ({ encoding?: BufferEncoding; flag?: string } & Abortable)
    | undefined
    | null,
): Effect.Effect<FS, ReadFileError, string | Buffer> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, ReadFileError, string | Buffer>(resume => {
        fs.readFile(path, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new ReadFileError({ path, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const mkdir = (
  file: NFS.PathLike,
  options: NFS.MakeDirectoryOptions & {
    recursive: true;
  } = {
    // TODO!!!! Check mode
    mode: 0,
    recursive: true,
  },
): Effect.Effect<FS, MkDirError, string | undefined> =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, MkDirError, string | undefined>(resume => {
        fs.mkdir(file, options, (error, data) => {
          if (error) {
            resume(Effect.fail(new MkDirError({ file, options, error })));
          } else {
            resume(Effect.succeed(data));
          }
        });
      }),
    ),
  );

export const stat = (path: string) =>
  pipe(
    FS,
    Effect.flatMap(fs =>
      Effect.async<FS, StatError, NFS.Stats>(resume =>
        { fs.stat(path, (error, stats) => {
          if (error) {
            resume(Effect.fail(new StatError({ path, error })));
          } else {
            resume(Effect.succeed(stats));
          }
        }); },
      ),
    ),
  );

export const isFileOrDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile() || _.isDirectory()),
  );

export const isFile = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile()),
  );

export const isDirectory = (
  pathLike: string,
): Effect.Effect<FS, StatError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isDirectory()),
  );

// export const open = (
//   path: PathLike,
//   flags: OpenMode | undefined,
//   mode: Mode | undefined | null
// ): Effect.Effect<FS, OpenError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, OpenError, number>(resume =>
//         fs.open(path, flags, mode, (error, fd) => {
//           if (error) {
//             resume(Effect.fail(new OpenError({ path, flags, mode, error })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );

// export const close = (fd: number): Effect.Effect<FS, CloseError, number> =>
//   pipe(
//     FS,
//     Effect.flatMap(fs =>
//       Effect.async<FS, CloseError, number>(resume =>
//         fs.close(fd, error => {
//           if (error) {
//             resume(Effect.fail(new CloseError({ error, fd })));
//           } else {
//             resume(Effect.succeed(fd));
//           }
//         })
//       )
//     )
//   );
