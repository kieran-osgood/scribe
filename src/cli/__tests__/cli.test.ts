import { _Cli } from '@scribe/cli';
import { PassThrough } from 'stream';
import getStream from 'get-stream';
import { BaseContext } from 'clipanion/lib/advanced/Cli';
import memfs from 'memfs';
import * as fs from 'fs';
import path from 'path';

vi.mock('simple-git', () => ({
  default: () => ({
    status(_: unknown, cb: (_: null, opts: { isClean(): boolean }) => void) {
      cb(null, { isClean: () => true });
    },
  }),
}));

// vi.mock('cosmiconfig', () => ({
//   cosmiconfig: () => ({
//     load() {
//       return Promise.resolve({
//         config: config,
//         filePath: 'scribe.config.ts',
//         isEmpty: false,
//       });
//     },
//   }),
// }));
//
beforeEach(() => {
  memfs.vol.mkdirSync(process.cwd(), { recursive: true });
  memfs.vol.mkdirSync(path.join(process.cwd(), 'test'), { recursive: true });

  memfs.vol.writeFileSync(
    path.join(process.cwd(), 'scribe.config.ts'),
    fs.readFileSync(
      path.join(process.cwd(), 'test/config/good-scribe.config.ts')
    )
  );
  memfs.vol.writeFileSync(
    path.join(process.cwd(), 'screen.scribe'),
    fs.readFileSync('test/screen.scribe')
  );
  memfs.vol.writeFileSync(
    path.join(process.cwd(), 'screen.test.scribe'),
    fs.readFileSync('test/screen.test.scribe')
  );
});

afterEach(() => memfs.vol.reset());

beforeAll(() => {
  process.env.NODE_ENV = 'production';
});

afterAll(() => {
  process.env.NODE_ENV = 'test';
});

vitest.mock('process', () => ({
  cwd: () => 'mockdir',
}));

function createCtx(): BaseContext {
  return {
    stdout: new PassThrough(),
    stdin: new PassThrough(),
    env: process.env,
    stderr: process.stderr,
    colorDepth: 0,
  };
}

describe('_Cli', () => {
  // it('should print default usage info', async () => {
  //   const passThrough = new PassThrough();
  //   await _Cli([''], {
  //     stdout: passThrough,
  //   });
  //   passThrough.end();
  //
  //   expect(await getStream(passThrough)).toMatchSnapshot();
  // });
  //
  // it('should print help info', async () => {
  //   const passThrough = new PassThrough();
  //   await _Cli(['--help'], {
  //     stdout: passThrough,
  //   });
  //   passThrough.end();
  //
  //   expect(await getStream(passThrough)).toMatchSnapshot();
  // });

  // it.skip('should accept name as argument and prompt for template', async () => {
  //   const stdout = new PassThrough();
  //   await _Cli(['--name=Login --config=./test/config/good-scribe.config.ts'], {
  //     stdout,
  //   });
  //   stdout.end();
  //
  //   const result = await getStream(stdout);
  //   console.log(result);
  // });
  //
  // it.skip('should accept template as argument and prompt for name', async () => {
  //   const stdout = new PassThrough();
  //   await _Cli(['--template=screen'], {
  //     stdout,
  //   });
  //   stdout.end();
  //
  //   expect(await getStream(stdout)).toMatchSnapshot();
  // });

  it.only('should accept template & fileName and complete', async () => {
    const ctx = createCtx();
    const args = [
      '--config=./test/config/good-scribe.config.ts',
      '--template=screen',
      '--name=Login',
    ];

    await _Cli(args, ctx);
    ctx.stdout.end();

    const result = await getStream(ctx.stdout);
    expect(result).toMatchInlineSnapshot(`
  "[32m‚úÖ  Success!
  [39mOutput files:
  - /Users/kieranosgood/WebstormProjects/scribe/examples/src/screens/Login.ts
  - /Users/kieranosgood/WebstormProjects/scribe/examples/src/screens/Login.test.ts
  Complete"
`);
  });

  // or no git repo?
  // it.only('should log dirty tree if user has changes', async () => {
  //   const passThrough = new PassThrough();
  //   const stdin = new PassThrough();
  //   process.env.NODE_ENV = 'production';
  //   await _Cli(['--name=Login --template=screen'], {
  //     stdout: passThrough,
  //     stdin,
  //   });
  //   passThrough.end();
  //
  //   expect(await getStream(passThrough)).toMatchInlineSnapshot('"‚ö†Ô∏è Working directory not clean"');
  // });
});
