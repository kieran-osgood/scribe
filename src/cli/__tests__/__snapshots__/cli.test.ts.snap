// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`_Cli > Init Command > should fail if file doesn't exist 2`] = `
"import type { ScribeConfig } from '@scribe/config';

const config = {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['./src/test-fixtures'],
  },
  templates: {
    component: {
      outputs: [
        {
          templateFileKey: 'component',
          output: {
            directory: 'examples/src/components',
            fileName: '{{Name}}.ts',
          },
        },
      ],
    },
    screen: {
      outputs: [
        {
          templateFileKey: 'screen',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.ts',
          },
        },
        {
          templateFileKey: 'screen.test',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.test.ts',
          },
        },
      ],
    },
  },
} satisfies ScribeConfig;

export default config;
"
`;

exports[`_Cli > Init Command > should write base config file 2`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;
